### PR TITLE
Feature/file restriction

### PR DESCRIPTION
**Pull request by franciscog-kainos**
### Work checklist

- [ ] Tests added where applicable
- [ ] UI changes look good on mobile
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are commited to keeping commit history clean, consistent and linear. To achive this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/controllers/EvidenceDownloadController.ts [hunk 1]
```diff
@@ -13,16 +18,22 @@ import { DOWNLOAD_FILE_PAGE_URI } from 'app/utils/Paths';
...
     }
 
     @httpGet('/prompt/:fileId')
-    public renderPrompt(@requestParam('fileId') fileId: string): void {
-        this.httpContext.response.render(template, { fileId });
+    public async renderPrompt(@requestParam('fileId') fileId: string): Promise<void> {
+        return this.render(template, { downloadPath: this.getDownloadPath(this.httpContext.request, fileId) });
     }
 
     @httpGet('/data/:fileId/download')
```
Comment (n-r-b) [inline src/controllers/EvidenceDownloadController.ts:39]: Are the newly added parameters on the download link read by this endpoint at any point? or are they only used by the LoadAppealMiddleware?

File: test/controllers/EvidenceDownloadController.test.ts [hunk 2]
```diff
@@ -1,167 +1,289 @@
...
+                av_status: 'clean',
+                content_type: DEFAULT_ATTACHMENTS[0].contentType,
+                id: DEFAULT_ATTACHMENTS[0].id,
+                name: DEFAULT_ATTACHMENTS[0].name,
+                size: DEFAULT_ATTACHMENTS[0].size
+            };
 
-        const fakeFileTransferProxy =
-            fileTransferServiceProxy(Promise.resolve(readable), metadataClean);
+            it('should render the prompt page correctly', async () => {
```
Comment (n-r-b) [inline test/controllers/EvidenceDownloadController.test.ts:121]: As there are two GET endpoints in this controller, I feel we should split the group of tests into two

File: src/views/download-file.njk [hunk 3]
```diff
@@ -8,7 +8,7 @@
     <div class="govuk-grid-column-two-thirds">
       <h1 class="govuk-heading-xl">Your document download will start soon</h1>
       <p class="govuk-body-l">
-        <a href="{{ paths.DOWNLOAD_FILE_PAGE_URI }}/data/{{ fileId }}/download" id="downloadLink" class="govuk-link">Click here to download your document</a>
+        <a href="{{ paths.DOWNLOAD_FILE_PAGE_URI }}/data/{{ downloadPath | safe }}" id="downloadLink" class="govuk-link">Click here to download your document</a>
```
[the remark this answers is not in the document]
Comment (franciscog-kainos) [reply src/views/download-file.njk:11]: it doesn't escape & and ?

File: src/controllers/EvidenceDownloadController.ts [hunk 4]
```diff
@@ -34,21 +45,44 @@ export class EvidenceDownloadController extends BaseAsyncHttpController {
...
+        if (!appealId && !companyNumber) {
+            return '';
+        }
+        if (appealId && !companyNumber) {
+            return `?a=${appealId}`;
+        }
+        if (!appealId) {
+            return `?c=${companyNumber}`;
+        }
+        return `${fileId}/download?a=${appealId}&c=${companyNumber}`;
```
[the remark this answers is not in the document]
Comment (franciscog-kainos) [reply src/controllers/EvidenceDownloadController.ts:82]: This scenario would have been caught by the LoadAppealMiddleware. So it's probably best to be consistent and throw an error if there is no company number

File: src/controllers/EvidenceDownloadController.ts [hunk 5]
```diff
@@ -34,21 +45,44 @@ export class EvidenceDownloadController extends BaseAsyncHttpController {
...
+    }
 
-        this.httpContext.response.setHeader('content-disposition', `attachment; filename=${metadata.name}`);
-        return await this.pipeDataIntoStream(readable, this.httpContext.response);
+    private getDownloadPath(req: Request, fileId: string): string {
+
+        const appealId = req.query[APPEAL_ID_QUERY_KEY];
+        const companyNumber = req.query[COMPANY_NUMBER_QUERY_KEY];
+
+        if (!appealId && !companyNumber) {
```
[the remark this answers is not in the document]
Comment (franciscog-kainos) [reply src/controllers/EvidenceDownloadController.ts:73]: These tests are in LoadAppealMiddleware. This logic should be as simple as throwing an error here. Perhaps we could add some tests in case the user decides to manipulate the urls

File: src/controllers/EvidenceDownloadController.ts [hunk 6]
```diff
@@ -13,16 +18,22 @@ import { DOWNLOAD_FILE_PAGE_URI } from 'app/utils/Paths';
...
 
     constructor(@inject(FileTransferService) private readonly fileTransferService: FileTransferService) {
         super();
     }
 
     @httpGet('/prompt/:fileId')
-    public renderPrompt(@requestParam('fileId') fileId: string): void {
-        this.httpContext.response.render(template, { fileId });
+    public async renderPrompt(@requestParam('fileId') fileId: string): Promise<void> {
+        return this.render(template, { downloadPath: this.getDownloadPath(this.httpContext.request, fileId) });
```
[the remark this answers is not in the document]
Comment (franciscog-kainos) [reply src/controllers/EvidenceDownloadController.ts:36]: This scenario will be caught by the FileRestrictionsMiddleware

File: src/controllers/EvidenceDownloadController.ts (hunk 1 above)
Comment (franciscog-kainos) [reply src/controllers/EvidenceDownloadController.ts:39]: They are ready by the `this.getDownloadPath` function

File: src/controllers/EvidenceDownloadController.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (franciscog-kainos) [reply src/controllers/EvidenceDownloadController.ts:82]: If there is no appeal Id the route won't be matched

File: src/controllers/EvidenceDownloadController.ts (hunk 1 above)
Comment (n-r-b) [reply src/controllers/EvidenceDownloadController.ts:39]: Ok, so what's the reason for generating a download path with the appeal ID and companyNumber if they parameters are then not read by the download endpoint? i'm assuming by this point, the appeal has already been loaded and is in the session. Or am i missing something here?

File: src/controllers/EvidenceDownloadController.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (n-r-b) [reply src/controllers/EvidenceDownloadController.ts:73]: Yeah i did see that the LoadApealMiddleware checks for these parameters and there are tests covering this. I agree with tests covering this logic incase of manipulation

File: src/controllers/EvidenceDownloadController.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (n-r-b) [reply src/controllers/EvidenceDownloadController.ts:78]: I can't see where we are generating this url and rendering the page with it: `{{ paths.DOWNLOAD_FILE_PAGE_URI }}/data/{{ fileId }}/download`, that's the reason I asked

File: src/controllers/EvidenceDownloadController.ts (hunk 1 above)
Comment (franciscog-kainos) [reply src/controllers/EvidenceDownloadController.ts:40]: They are used to create the download link in the prompt page

File: test/ApplicationFactory.ts [hunk 7]
```diff
@@ -24,14 +24,17 @@ export const createAppConfigurable = (configureBindings: (container: Container)
 };
 
 export const createApp = (data?: Partial<ApplicationData>,
-                          // tslint:disable-next-line:no-empty
-                          configureBindings: (container: Container) => void = () => {}) =>
+    // tslint:disable-next-line:no-empty
+    configureBindings: (container: Container) => void = () => { },
+    // tslint:disable-next-line:no-empty
```
Comment (akaJSaunders) [inline test/ApplicationFactory.ts:29]: is this tslint:disable-next-line:no-empty being used?

File: test/ApplicationFactory.ts [hunk 8]
```diff
@@ -24,14 +24,17 @@ export const createAppConfigurable = (configureBindings: (container: Container)
 };
 
 export const createApp = (data?: Partial<ApplicationData>,
-                          // tslint:disable-next-line:no-empty
-                          configureBindings: (container: Container) => void = () => {}) =>
+    // tslint:disable-next-line:no-empty
+    configureBindings: (container: Container) => void = () => { },
```
Comment (akaJSaunders) [inline test/ApplicationFactory.ts:28]: can we remove the space in between those two braces 🏴󠁧󠁢󠁷󠁬󠁳󠁿

Comment (akaJSaunders) [reply test/ApplicationFactory.ts:28]:
```suggestion
    configureBindings: (container: Container) => void = () => {},
```

File: test/ApplicationFactory.ts (hunk 7 above)
Comment (akaJSaunders) [reply test/ApplicationFactory.ts:29]:
```suggestion
```

File: src/controllers/EvidenceDownloadController.ts (hunk 1 above)
Comment (n-r-b) [reply src/controllers/EvidenceDownloadController.ts:40]: discussed

File: src/controllers/CheckYourAppealController.ts [hunk 9]
```diff
@@ -63,6 +63,6 @@ export class CheckYourAppealController extends SafeNavigationBaseController<any>
     }
 
     protected prepareViewModelFromAppeal(appeal: Appeal): any {
-        return appeal;
+        return { ...appeal, companyNumber: appeal.penaltyIdentifier?.companyNumber };
```
Comment (n-r-b) [inline src/controllers/CheckYourAppealController.ts:66]: as discussed - please revert this change

File: src/controllers/EvidenceDownloadController.ts [hunk 10]
```diff
@@ -13,16 +18,23 @@ import { DOWNLOAD_FILE_PAGE_URI } from 'app/utils/Paths';
 const template = 'download-file';
 const errorCustomTemplate = 'error-custom';
 
-@controller(DOWNLOAD_FILE_PAGE_URI, FileTransferFeatureMiddleware)
+@controller(DOWNLOAD_FILE_PAGE_URI,
+    SessionMiddleware,
+    FileRestrictionsAuthMiddleware,
```
Comment (n-r-b) [inline src/controllers/EvidenceDownloadController.ts:23]: Let's say we have disabled the file transfer feature, would the Auth Middleware specific to File Restrictions cause any issues?

File: src/models/CreatedBy.ts [hunk 11]
```diff
@@ -1,3 +1,3 @@
 export interface CreatedBy {
-    _id: string;
+    id: string;
```
Comment (n-r-b) [inline src/models/CreatedBy.ts:2]: does this mean we can remove the  `@JsonAlias({"_id"})` from the api?

File: src/controllers/EvidenceDownloadController.ts (hunk 10 above)
Comment (franciscog-kainos) [reply src/controllers/EvidenceDownloadController.ts:23]: The user would log in as usual but then it will just kick you to the penalty details screen

File: src/controllers/EvidenceDownloadController.ts [hunk 12]
```diff
@@ -34,21 +46,39 @@ export class EvidenceDownloadController extends BaseAsyncHttpController {
...
 
-        this.httpContext.response.setHeader('content-disposition', `attachment; filename=${metadata.name}`);
-        return await this.pipeDataIntoStream(readable, this.httpContext.response);
+    private getDownloadPath(req: Request, fileId: string): string {
+
+        const appealId = req.query[APPEAL_ID_QUERY_KEY];
+        const companyNumber = req.query[COMPANY_NUMBER_QUERY_KEY];
+
+        if (!appealId || !companyNumber) {
+            throw Error('Expected appealId and company number as query parameters but none found');
```
Comment (n-r-b) [inline src/controllers/EvidenceDownloadController.ts:75]: Should this error message be: `'Expected **either** appealId **or** company number as query parameters but none found'`

File: src/models/CreatedBy.ts (hunk 11 above)
Comment (franciscog-kainos) [reply src/models/CreatedBy.ts:2]: yep :)

File: src/controllers/EvidenceDownloadController.ts (hunk 12 above)
Comment (n-r-b) [reply src/controllers/EvidenceDownloadController.ts:75]: or should this check `(!appealId && !companyNumber)`?

Comment (franciscog-kainos) [reply src/controllers/EvidenceDownloadController.ts:75]: either not being present is correct

File: src/middleware/LoadAppealMiddleware.ts [hunk 13]
```diff
@@ -21,41 +22,41 @@ export class LoadAppealMiddleware extends BaseMiddleware {
     constructor(@inject(AppealsService) private readonly appealsService: AppealsService) {
         super();
     }
-    // @ts-ignore
-    public async handler(req: Request, res: Response, next: NextFunction): Promise<void> {
+    public handler: RequestHandler = makeAsyncRequestHandler(
+        // @ts-ignore
+        async (req: Request, res: Response, next: NextFunction): Promise<void> => {
+
+            const companyNumber = req.query[COMPANY_NUMBER_QUERY_KEY] as string;
```
Comment (n-r-b) [inline src/middleware/LoadAppealMiddleware.ts:29]: if there is no company number or appealId?

Comment (franciscog-kainos) [reply src/middleware/LoadAppealMiddleware.ts:29]: Check below
```
 try {
            new SchemaValidator(companyNumberSchema).validate(companyNumber);
       } catch (err) {
            throw new Error('Tried to load appeal from an invalid company number');
        }
// and
 if (appealId && !applicationData.appeal) 
```

Comment (franciscog-kainos) [reply src/middleware/LoadAppealMiddleware.ts:29]: Those cases are covered :). If there is no appealId then the file restrictions middleware will take it as the user is still completing the appeal

Comment (n-r-b) [reply src/middleware/LoadAppealMiddleware.ts:29]: ah i see, thanks for clarifying 👍
